src/main/host/syscall/socket.rs

```diff
@@ -0,0 +1,578 @@
+        );
+        UnixSocketFile::connect(
+            &socket_2,
+            unnamed_sock_addr,
+            Arc::clone(socket_1.borrow().recv_buffer()),
+            event_queue,
+        );
+
+        socket_1
+            .borrow_mut()
+            .set_bound_address(Some(unnamed_sock_addr))
+            .unwrap();
+        socket_2
+            .borrow_mut()
+            .set_bound_address(Some(unnamed_sock_addr))
+            .unwrap();
+    });
+
+    // file descriptors for the sockets
+    let mut desc_1 = Descriptor::new(PosixFile::Socket(SocketFile::Unix(socket_1)));
+    let mut desc_2 = Descriptor::new(PosixFile::Socket(SocketFile::Unix(socket_2)));
+
+    // set the file descriptor flags
+    desc_1.set_flags(descriptor_flags);
+    desc_2.set_flags(descriptor_flags);
+
+    // register the file descriptors
+    let fd_1 = ctx
+        .process
+        .register_descriptor(CompatDescriptor::New(desc_1));
+    let fd_2 = ctx
+        .process
+        .register_descriptor(CompatDescriptor::New(desc_2));
+
+    // try to write them to the caller
+    let fds = [i32::try_from(fd_1).unwrap(), i32::try_from(fd_2).unwrap()];
+    let write_res = ctx
+        .process
+        .memory_mut()
+        .copy_to_ptr(TypedPluginPtr::new::<libc::c_int>(fd_ptr.into(), 2), &fds);
+
+    // clean up in case of error
+    match write_res {
+        Ok(_) => Ok(0.into()),
+        Err(e) => {
+            EventQueue::queue_and_run(|event_queue| {
+                // ignore any errors when closing
+                ctx.process
+                    .deregister_descriptor(fd_1)
+                    .unwrap()
+                    .close(ctx.host.chost(), event_queue);
+                ctx.process
+                    .deregister_descriptor(fd_2)
+                    .unwrap()
+                    .close(ctx.host.chost(), event_queue);
+            });
+            Err(e.into())
+        }
+    }
+}
+
+/// Returns a nix socket address object where only the family is set.
+fn empty_sockaddr(family: nix::sys::socket::AddressFamily) -> nix::sys::socket::SockAddr {
+    let family = family as libc::sa_family_t;
+    let mut addr: nix::sys::socket::sockaddr_storage = unsafe { std::mem::zeroed() };
+    addr.ss_family = family;
+    // the size of ss_family will be 2 bytes on linux
+    nix::sys::socket::sockaddr_storage_to_addr(&addr, 2).unwrap()
+}
+
+/// Copy the socket address to the plugin. Will return an error if either the address or address
+/// length pointers are NULL. The plugin's address length will be updated to store the size of the
+/// socket address, even if greater than the provided buffer size. If the address is `None`, the
+/// plugin's address length will be set to 0.
+fn write_sockaddr(
+    process: &mut Process,
+    addr: Option<nix::sys::socket::SockAddr>,
+    plugin_addr: PluginPtr,
+    plugin_addr_len: TypedPluginPtr<libc::socklen_t>,
+) -> Result<(), SyscallError> {
+    if plugin_addr.is_null() || plugin_addr_len.is_null() {
```

Comment:
Redundant; the MemoryManager will return EFAULT for NULL pointers (assuming the usual case that they're actually inaccessible)

Reply:
This is here to avoid unexpected behaviour when one plugin ptr is null and the other isn't. For example if `plugin_addr` is null and `plugin_addr_len` isn't null, this function would modify the value at `plugin_addr_len` before trying to write to `plugin_addr` and returning `EFAULT`.

I added a comment here to explain the purpose of this check.

Reply:
Removed the early exit.